src/ring_theory/etale.lean

```diff
@@ -300,4 +313,125 @@ formally_etale.iff_unramified_and_smooth.mpr ⟨infer_instance, infer_instance
 
 end base_change
 
+section localization
+
+variables {R S Rₘ Sₘ : Type u} [comm_ring R] [comm_ring S] [comm_ring Rₘ] [comm_ring Sₘ]
+variables (M : submonoid R)
+variables [algebra R S] [algebra R Sₘ] [algebra S Sₘ] [algebra R Rₘ] [algebra Rₘ Sₘ]
+variables [is_scalar_tower R Rₘ Sₘ] [is_scalar_tower R S Sₘ]
+variables [is_localization M Rₘ] [is_localization (M.map (algebra_map R S)) Sₘ]
+local attribute [elab_as_eliminator] ideal.is_nilpotent.induction_on
+
+lemma is_unit.lift_of_nilpotent (x : R) {I : ideal R} (hI : is_nilpotent I)
+  (e : is_unit (ideal.quotient.mk I x)) : is_unit x :=
+begin
+  revert x e,
+  apply ideal.is_nilpotent.induction_on I hI; clear hI I,
+  swap,
+  { introv e h₁ h₂ h₃,
+    apply h₁,
+    apply h₂,
+    exactI h₃.map ((double_quot.quot_quot_equiv_quot_sup I J).trans
+      (ideal.quot_equiv_of_eq (sup_eq_right.mpr e))).symm.to_ring_hom },
+  { introv e H,
+    resetI,
+    obtain ⟨y, hy⟩ := ideal.quotient.mk_surjective (↑(H.unit⁻¹) : S ⧸ I),
+    have : ideal.quotient.mk I (x * y) = ideal.quotient.mk I 1,
+    { rw [map_one, map_mul, hy, is_unit.mul_coe_inv] },
+    rw ideal.quotient.eq at this,
+    have : (x * y - 1) ^ 2 = 0,
+    { rw [← ideal.mem_bot, ← e], exact ideal.pow_mem_pow this _ },
+    have : x * (y * (2 - x * y)) = 1,
+    { rw [eq_comm, ← sub_eq_zero, ← this], ring },
+    exact is_unit_of_mul_eq_one _ _ this }
+end
+
+include M
+
+lemma formally_smooth.to_localization : formally_smooth R Rₘ :=
```

Comment:
I don't understand this name. Shouldn't it be `is_localization.formally_smooth` or something like that?